plugins/LocalContainerProvider/LocalContainerProvider.py

```diff
@@ -0,0 +1,192 @@
+# Copyright (c) 2017 Ultimaker B.V.
+# Uranium is released under the terms of the LGPLv3 or higher.
+
+import os #For getting the IDs from a filename.
+import pickle #For caching definitions.
+import re #To detect back-up files in the ".../old/#/..." folders.
+from typing import Any, Dict, Iterable, Optional
+
+from UM.Application import Application #To get the current version for finding the cache directory.
+from UM.Settings.ContainerRegistry import ContainerRegistry #To get the resource types for containers.
+from UM.Logger import Logger
+from UM.MimeTypeDatabase import MimeTypeDatabase, MimeType #To get the type of container we're loading.
+from UM.Settings.ContainerProvider import ContainerProvider #The class we're implementing.
+from UM.Settings.DefinitionContainer import DefinitionContainer #To check if we need to cache this container.
+from UM.Resources import Resources
+
+MYPY = False
+if MYPY: #Things to import for type checking only.
+    from UM.Settings.Interfaces import ContainerInterface
+
+##  Provides containers from the local installation.
+class LocalContainerProvider(ContainerProvider):
+    ##  Creates the local container provider.
+    #
+    #   This creates a cache which translates container IDs to their file names.
+    def __init__(self):
+        super().__init__()
+
+        self._id_to_path = {} # type: Dict[str, str] #Translates container IDs to the path to where the file is located.
+        self._id_to_mime = {} # type: Dict[str, MimeType] #Translates container IDs to their MIME type.
+
+    ##  Gets the IDs of all local containers.
+    #
+    #   \return A sequence of all container IDs.
+    def getAllIds(self) -> Iterable[str]:
+        if not self._id_to_path:
+            self._updatePathCache()
+        return self._id_to_path.keys()
+
+    def loadContainer(self, container_id: str) -> "ContainerInterface":
+        Logger.log("d", "Loading container {container_id}".format(container_id = container_id))
+        container_class = ContainerRegistry.mime_type_map[self._id_to_mime[container_id].name]
+        if issubclass(container_class, DefinitionContainer): #We may need to load these from the definition cache.
+            container = self._loadCachedDefinition(container_id)
```

Comment:
Is it possible to try this first? After retrieving we can do the class test. Is the class test necessary?

Reply:
Getting the cache first would work, but it would cause us to check for existence of the cache file, which requires a system call and then hard disk access. That's much slower than checking a property of an instance, so I think it's best to check this first.